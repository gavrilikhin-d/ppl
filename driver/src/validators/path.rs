use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
