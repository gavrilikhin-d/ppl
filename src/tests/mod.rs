use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
