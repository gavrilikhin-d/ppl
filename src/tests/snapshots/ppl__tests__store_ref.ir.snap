---
source: src/tests/mod.rs
expression: ir
---
; ModuleID = 'main'
source_filename = "src/main.ppl"

%"Type<String>" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }
%"Type<MemoryAddress>" = type { %String, %Integer }
%"Type<I32>" = type { %String, %Integer }
%"Type<F64>" = type { %String, %Integer }
%RefWrapper = type { ptr }

@"Type<String>" = private global %"Type<String>" zeroinitializer
@0 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@"Type<MemoryAddress>" = private global %"Type<MemoryAddress>" zeroinitializer
@1 = private unnamed_addr constant [14 x i8] c"MemoryAddress\00", align 1
@"Type<I32>" = private global %"Type<I32>" zeroinitializer
@2 = private unnamed_addr constant [4 x i8] c"I32\00", align 1
@"Type<F64>" = private global %"Type<F64>" zeroinitializer
@3 = private unnamed_addr constant [4 x i8] c"F64\00", align 1
@x = global %Integer zeroinitializer
@y = global %RefWrapper zeroinitializer

define private void @initialize() !dbg !3 {
  %1 = alloca %"Type<String>", align 8, !dbg !7
  %"Type<String>.name" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 6), !dbg !8
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !8
  %"Type<String>.size" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !8
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !8
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define private void @initialize.1() !dbg !9 {
  %1 = alloca %"Type<MemoryAddress>", align 8, !dbg !10
  %"Type<MemoryAddress>.name" = getelementptr inbounds %"Type<MemoryAddress>", ptr %1, i32 0, i32 0, !dbg !10
  %2 = call %String @string_from_c_string_and_length(ptr @1, i64 13), !dbg !11
  store %String %2, ptr %"Type<MemoryAddress>.name", align 8, !dbg !11
  %"Type<MemoryAddress>.size" = getelementptr inbounds %"Type<MemoryAddress>", ptr %1, i32 0, i32 1, !dbg !11
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !11
  store %Integer %3, ptr %"Type<MemoryAddress>.size", align 8, !dbg !11
  %4 = load %"Type<MemoryAddress>", ptr %1, align 8, !dbg !11
  store %"Type<MemoryAddress>" %4, ptr @"Type<MemoryAddress>", align 8, !dbg !11
  br label %return, !dbg !11

return:                                           ; preds = %0
  ret void
}

define private void @initialize.2() !dbg !12 {
  %1 = alloca %"Type<I32>", align 8, !dbg !13
  %"Type<I32>.name" = getelementptr inbounds %"Type<I32>", ptr %1, i32 0, i32 0, !dbg !13
  %2 = call %String @string_from_c_string_and_length(ptr @2, i64 3), !dbg !14
  store %String %2, ptr %"Type<I32>.name", align 8, !dbg !14
  %"Type<I32>.size" = getelementptr inbounds %"Type<I32>", ptr %1, i32 0, i32 1, !dbg !14
  %3 = call %Integer @integer_from_i64(i64 4), !dbg !14
  store %Integer %3, ptr %"Type<I32>.size", align 8, !dbg !14
  %4 = load %"Type<I32>", ptr %1, align 8, !dbg !14
  store %"Type<I32>" %4, ptr @"Type<I32>", align 8, !dbg !14
  br label %return, !dbg !14

return:                                           ; preds = %0
  ret void
}

define private void @initialize.3() !dbg !15 {
  %1 = alloca %"Type<F64>", align 8, !dbg !16
  %"Type<F64>.name" = getelementptr inbounds %"Type<F64>", ptr %1, i32 0, i32 0, !dbg !16
  %2 = call %String @string_from_c_string_and_length(ptr @3, i64 3), !dbg !17
  store %String %2, ptr %"Type<F64>.name", align 8, !dbg !17
  %"Type<F64>.size" = getelementptr inbounds %"Type<F64>", ptr %1, i32 0, i32 1, !dbg !17
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !17
  store %Integer %3, ptr %"Type<F64>.size", align 8, !dbg !17
  %4 = load %"Type<F64>", ptr %1, align 8, !dbg !17
  store %"Type<F64>" %4, ptr @"Type<F64>", align 8, !dbg !17
  br label %return, !dbg !17

return:                                           ; preds = %0
  ret void
}

define void @main.execute() !dbg !18 {
  call void @initialize(), !dbg !19
  call void @initialize.1(), !dbg !19
  call void @initialize.2(), !dbg !20
  call void @initialize.3(), !dbg !21
  call void @initialize.4(), !dbg !22
  call void @initialize.5(), !dbg !23
  %1 = load ptr, ptr @y, align 8, !dbg !24
  %2 = call %Integer @clone_integer(ptr %1), !dbg !24
  call void @"println <:Integer>"(%Integer %2), !dbg !24
  call void @destroy_integer(ptr @x), !dbg !25
  br label %return, !dbg !25

return:                                           ; preds = %0
  ret void
}

define %RefWrapper @"wrap <:Reference<Integer>>"(ptr %0) !dbg !26 {
  %return_value = alloca %RefWrapper, align 8
  %x = alloca ptr, align 8
  store ptr %0, ptr %x, align 8
  %2 = alloca %RefWrapper, align 8, !dbg !27
  %RefWrapper.x = getelementptr inbounds %RefWrapper, ptr %2, i32 0, i32 0, !dbg !27
  %3 = call %Integer @clone_integer(ptr %x), !dbg !28
  store %Integer %3, ptr %RefWrapper.x, align 8, !dbg !28
  %4 = load %RefWrapper, ptr %2, align 8, !dbg !28
  %"$tmp@56" = alloca %RefWrapper, align 8, !dbg !28
  store %RefWrapper %4, ptr %"$tmp@56", align 8, !dbg !28
  %5 = load %RefWrapper, ptr %"$tmp@56", align 8, !dbg !27
  store %RefWrapper %5, ptr %return_value, align 8, !dbg !27
  br label %return, !dbg !27

return:                                           ; preds = %1
  %6 = load %RefWrapper, ptr %return_value, align 8
  ret %RefWrapper %6
}

declare %Integer @clone_integer(ptr)

define private void @initialize.4() !dbg !29 {
  %1 = call %Integer @integer_from_i64(i64 5), !dbg !30
  store %Integer %1, ptr @x, align 8, !dbg !30
  br label %return, !dbg !30

return:                                           ; preds = %0
  ret void
}

define private void @initialize.5() !dbg !31 {
  %1 = call %RefWrapper @"wrap <:Reference<Integer>>"(ptr @x), !dbg !32
  store %RefWrapper %1, ptr @y, align 8, !dbg !32
  br label %return, !dbg !32

return:                                           ; preds = %0
  ret void
}

define private void @"println <:Integer>"(%Integer %0) !dbg !33 {
  %x = alloca %Integer, align 8
  store %Integer %0, ptr %x, align 8
  %2 = load %Integer, ptr %x, align 8, !dbg !34
  %3 = call %String @integer_as_string(%Integer %2), !dbg !34
  call void @"println <:String>"(%String %3), !dbg !34
  br label %return, !dbg !34

return:                                           ; preds = %1
  ret void
}

declare void @"println <:String>"(%String)

declare %String @integer_as_string(%Integer)

declare void @destroy_integer(ptr)

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "src/main.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 7, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 7, column: 11, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !2, file: !2, line: 7, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 7, column: 11, scope: !9)
!11 = !DILocation(line: 0, scope: !9)
!12 = distinct !DISubprogram(name: "initialize.2", linkageName: "initialize.2", scope: !2, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!13 = !DILocation(line: 3, column: 32, scope: !12)
!14 = !DILocation(line: 0, scope: !12)
!15 = distinct !DISubprogram(name: "initialize.3", linkageName: "initialize.3", scope: !2, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!16 = !DILocation(line: 3, column: 27, scope: !15)
!17 = !DILocation(line: 0, scope: !15)
!18 = distinct !DISubprogram(name: "main.execute", linkageName: "main.execute", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!19 = !DILocation(line: 7, column: 11, scope: !18)
!20 = !DILocation(line: 3, column: 32, scope: !18)
!21 = !DILocation(line: 3, column: 27, scope: !18)
!22 = !DILocation(line: 5, column: 8, scope: !18)
!23 = !DILocation(line: 6, column: 8, scope: !18)
!24 = !DILocation(line: 7, column: 8, scope: !18)
!25 = !DILocation(line: 5, scope: !18)
!26 = distinct !DISubprogram(name: "wrap <:Reference<Integer>>", linkageName: "wrap <:Reference<Integer>>", scope: !18, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!27 = !DILocation(line: 3, column: 25, scope: !26)
!28 = !DILocation(line: 3, column: 38, scope: !26)
!29 = distinct !DISubprogram(name: "initialize.4", linkageName: "initialize.4", scope: !18, file: !2, line: 5, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!30 = !DILocation(line: 5, column: 8, scope: !29)
!31 = distinct !DISubprogram(name: "initialize.5", linkageName: "initialize.5", scope: !18, file: !2, line: 6, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!32 = !DILocation(line: 6, column: 13, scope: !31)
!33 = distinct !DISubprogram(name: "println <:Integer>", linkageName: "println <:Integer>", scope: !18, file: !2, line: 7, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!34 = !DILocation(line: 7, column: 11, scope: !33)
